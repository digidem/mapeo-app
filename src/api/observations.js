// @flow
import fetch from 'fetch';
import { jsonRequest } from './base';

export interface CreateRequest {
  lat: number;
  lon: number;
}

export type UpdateRequest = {
  id: string
};

class Observation {
  static list = () =>
    jsonRequest({
      method: 'GET',
      route: '/observations'
    });

  static create = (observation: CreateRequest) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: {
        device_id: '1',
        ...observation
      }
    });

  static update = (observation: UpdateRequest) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: {
        ...observation
      }
    });
}

export default Observation;
