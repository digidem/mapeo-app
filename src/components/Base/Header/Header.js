import React from 'react';
import { StyleSheet, View, ActivityIndicator, Text } from 'react-native';
import I18n from 'react-native-i18n';
import { GREEN, WHITE, MAPEO_BLUE } from '../../../lib/styles';
import type { Resource } from '../../../types/redux';
import type { GPSState } from '../../../types/gps';

interface Props {
  leftIcon: any;
  rightIcon: any;
  style?: any;
  showTriangle?: boolean;
}

export interface StateProps {
  gps: Resource<GPSState>;
}

interface State {
  loading: boolean;
}

const styles = StyleSheet.create({
  headerBar: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    height: 60,
    paddingHorizontal: 15
  },
  gpsPill: {
    height: 35,
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, .8)',
    borderRadius: 50,
    paddingRight: 15,
    paddingLeft: 13
  },
  triangle: {
    position: 'absolute',
    width: 0,
    height: 0,
    top: 35,
    backgroundColor: 'transparent',
    borderStyle: 'solid',
    borderLeftWidth: 7,
    borderRightWidth: 7,
    borderBottomWidth: 7,
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'rgba(0, 0, 0, .8)',
    transform: [{ rotate: '180deg' }]
  }
});

I18n.fallbacks = true;
I18n.translations = {
  en: require('../../../translations/en'),
  es: require('../../../translations/es')
};

class Header extends React.PureComponent<Props & StateProps> {
  render() {
    const { leftIcon, rightIcon, style, showTriangle, gps } = this.props;
    const showLoading =
      gps &&
      (gps.status === 'Pending' || (gps.data && gps.data.accuracy > 100));

    let title = '';
    if (gps === undefined || gps.status === 'Failed') {
      title = I18n.t('gps.off');
    } else if (gps.status === 'Pending') {
      title = I18n.t('gps.loading');
    } else {
      title = `+/-  ${Math.round(gps.data.accuracy)}m`;
    }

    return (
      <View style={[styles.headerBar, style]}>
        {leftIcon}
        <View
          style={{
            justifyContent: 'center',
            alignItems: 'center'
          }}
        >
          <View style={styles.gpsPill}>
            {showLoading && <ActivityIndicator />}
            {!showLoading && (
              <View
                style={{
                  backgroundColor:
                    gps === undefined || gps.status === 'Failed'
                      ? 'lightgrey'
                      : GREEN,
                  height: 10,
                  width: 10,
                  borderRadius: 50
                }}
              />
            )}
            <Text style={{ color: WHITE, marginHorizontal: 20 }}>{title}</Text>
          </View>
          {showTriangle && <View style={styles.triangle} />}
        </View>
        {rightIcon}
      </View>
    );
  }
}

export default Header;
