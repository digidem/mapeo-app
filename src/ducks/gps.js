// @flow

import update from 'immutability-helper';
import { create } from '../lib/redux';
import {
  resourcePending,
  resourceSuccess,
  resourceFailed
} from '../lib/resource';

export const {
  type: GPS_PENDING,
  action: gpsPending,
  reducer: gpsPendingReducer
} = create('GPS_PENDING', {
  start: (state, meta) => {
    const newState = update(state, {
      gps: { $set: resourcePending(meta) }
    });

    return newState;
  }
});

export const { type: GPS_SET, action: gpsSet, reducer: gpsSetReducer } = create(
  'GPS_SET',
  {
    start: (state, meta) => {
      const newState = update(state, {
        gps: { $set: resourceSuccess(meta) }
      });

      return newState;
    }
  }
);

export const {
  type: GPS_FAILED,
  action: gpsFailed,
  reducer: gpsFailedReducer
} = create('GPS_FAILED', {
  start: (state, meta) => {
    const newState = update(state, {
      gps: { $set: resourceFailed(meta) }
    });

    return newState;
  }
});

export default [gpsPendingReducer, gpsSetReducer, gpsFailedReducer];
