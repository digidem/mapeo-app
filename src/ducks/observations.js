// @flow

import update from 'immutability-helper';
import {keyBy} from 'lodash';
import {create} from '../lib/redux';
import {createSelector} from '../lib/selector';
import type {StoreState} from '../types/redux';
import type {Observation} from '../types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const observations = keyBy(action.payload, 'id');

    if (state.observations) {
      Object.keys(state.observations).forEach(id => observations[id] = {
        ...observations[id],
        ...state.observations[id],
      });
    }

    const newState = update(state, {observations: {$set: observations}});

    return newState;
  }
});

export const {
  type: OBSERVATION_CREATE,
  action: observationCreate,
  reducer: observationCreateReducer
} = create('OBSERVATION_CREATE', {
  start: (state, action) => {
    const observation = action.meta;

    if (!observation.lat && !observation.lon && state.gps.data) {
      observation.lat = state.gps.data.latitude;
      observation.lon = state.gps.data.longitude;
    }

    const newState = update(state, {selectedObservation: {$set: observation}});

    return newState;
  },
  success: (state, action) => {
    const newState = update(state, {selectObservation: {$set: action.payload}});

    return newState;
  }
});

export const {
  type: OBSERVATION_UPDATE,
  action: observationUpdate,
  reducer: observationUpdateReducer
} = create('OBSERVATION_UPDATE', {
  start: (state, action) => {
    let newState;
    if (state.selectedObservation &&
        action.meta.id === state.selectedObservation.id) {
      newState = update(state, {
        selectedObservation: {
          $set: {...state.selectedObservation, ...action.meta},
        }
      });
    } else {
      newState = update(state, {
        observations: {
          [action.meta.id]: {
            $set: {...state.observations[action.meta.id], ...action.meta}
          }
        }
      });
    }

    return newState;
  }
});

export const {
  type: OBSERVATION_SELECT,
  action: observationSelect,
  reducer: observationSelectReducer
} = create('OBSERVATION_SELECT', {
  start: (state, action) => {
    if (action.meta === undefined && state.selectedObservation) {
      return update(state, {
        selectedObservation: {$set: action.meta},
        observations: {
          [state.selectedObservation.id]: {
            $set: {
              ...(state.observations[state.selectedObservation.id] || {}),
              ...state.selectedObservation,
            }
          }
        }
      });
    } else {
      return update(state, {selectedObservation: {$set: action.meta}});
    }
  }
});

export const selectObservation = createSelector(
    [(state: StoreState, id: string): Observation =>
         state.app.observations[id]],
    (observation: Observation): Observation => observation,
    (observation: Observation, id: string): string => id);

export default [
  observationListReducer, observationCreateReducer, observationUpdateReducer,
  observationSelectReducer
];
