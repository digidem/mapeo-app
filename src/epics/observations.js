// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import {
  OBSERVATION_LIST,
  observationList,
  OBSERVATION_CREATE,
  observationCreate,
  OBSERVATION_UPDATE,
  observationUpdate
} from '../ducks/observations';
import { Action } from '../types/redux';
import type { Observation as ObservationType } from '../types/observation';
import type { StoreState } from '../types/redux';
import Observation from '../api/observations';
import type { CreateRequest, UpdateRequest } from '../api/observations';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, ObservationType[]>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(action => action.status === 'Start')
    .flatMap(() =>
      Observation.list().map(observations => {
        console.log('RN - Observations List', observations);
        return observationList('', observations);
      })
    );

export const observationCreateEpic = (
  action$: ActionsObservable<Action<CreateRequest, ObservationType>>,
  store: StoreState
) =>
  action$
    .ofType(OBSERVATION_CREATE)
    .filter(action => action.status === 'Start')
    .flatMap(action =>
      Observation.create(action.meta).map(observation => {
        console.log('RN - Observation Created');
        return observationCreate(action.meta, observation);
      })
    );

// export const observationUpdateEpic = (
//   action$: ActionsObservable<Action<UpdateRequest, ObservationType>>,
//   store: StoreState
// ) =>
//   action$
//     .ofType(OBSERVATION_UPDATE)
//     .filter(action => action.status === 'Start')
//     .flatMap(action =>
//       Observation.update(action.meta).map(observation => {
//         console.log('RN - Observation Updated - ', action.meta, observation);
//         return observationUpdate(action.meta, observation);
//       })
//     );

export default [
  observationListEpic,
  observationCreateEpic
  // observationUpdateEpic
];
