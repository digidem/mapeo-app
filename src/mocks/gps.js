// @flow
import type { GPSState } from '../types/gps';
import {
  PERMISSION_DENIED,
  UNAVAILABLE,
  SEARCHING,
  LOW_ACCURACY,
  HIGH_ACCURACY
} from '../ducks/gps';

export const createGPSState = (partial: any): GPSState => ({
  status: HIGH_ACCURACY,
  coords: {
    longitude: 1,
    latitude: 1,
    accuracy: 50
  },
  timestamp: 0,

  ...partial
});
