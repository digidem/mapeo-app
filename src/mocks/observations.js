// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 1,
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: new Date('2018-02-25T20:23:00.000Z'),
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: require('../images/categories/category_18.png'),
  categoryId: 19,

  ...observation
});
