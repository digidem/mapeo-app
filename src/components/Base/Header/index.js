// @flow
import { connect } from 'react-redux';
import { StoreState } from '../../../types/redux';
import Header from './Header';
import type { StateProps } from './Header';

const mapStateToProps = (state: StoreState): StateProps => ({
  gps: state.gps
});

export default connect(mapStateToProps, null)(Header);
