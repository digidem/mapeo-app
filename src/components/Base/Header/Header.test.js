// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import Header from './Header';
import { createObservation } from '../../../mocks/observations';
import { createGPSState } from '../../../mocks/gps';

jest.mock('../../../ducks/gps', () => ({
  PERMISSION_DENIED: 'PERMISSION_DENIED',
  UNAVAILABLE: 'UNAVAILABLE',
  SEARCHING: 'SEARCHING',
  LOW_ACCURACY: 'LOW_ACCURACY',
  HIGH_ACCURACY: 'HIGH_ACCURACY'
}));

describe('Header tests', () => {
  const gps = createGPSState();

  test('snapshot without triangle prop', () => {
    const tree = renderer.create(<Header gps={gps} leftIcon="" rightIcon="" />);
    expect(tree).toMatchSnapshot();
  });

  test('snapshot with triangle prop', () => {
    const tree = renderer.create(
      <Header showTriangle gps={gps} leftIcon="" rightIcon="" />
    );
    expect(tree).toMatchSnapshot();
  });
});
