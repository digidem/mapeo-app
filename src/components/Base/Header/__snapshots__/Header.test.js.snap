// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests snapshot with triangle prop 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "paddingHorizontal": 15,
      },
      undefined,
    ]
  }
>
  
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, .8)",
          "borderRadius": 50,
          "flexDirection": "row",
          "height": 35,
          "paddingLeft": 13,
          "paddingRight": 15,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#7AFA4C",
            "borderRadius": 50,
            "height": 10,
            "width": 10,
          }
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "marginHorizontal": 20,
          }
        }
      >
        ± 50m
      </Text>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "transparent",
          "borderBottomColor": "rgba(0, 0, 0, .8)",
          "borderBottomWidth": 7,
          "borderLeftColor": "transparent",
          "borderLeftWidth": 7,
          "borderRightColor": "transparent",
          "borderRightWidth": 7,
          "borderStyle": "solid",
          "height": 0,
          "position": "absolute",
          "top": 35,
          "transform": Array [
            Object {
              "rotate": "180deg",
            },
          ],
          "width": 0,
        }
      }
    />
  </View>
  
</View>
`;

exports[`Header tests snapshot without triangle prop 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "space-between",
        "paddingHorizontal": 15,
      },
      undefined,
    ]
  }
>
  
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, .8)",
          "borderRadius": 50,
          "flexDirection": "row",
          "height": 35,
          "paddingLeft": 13,
          "paddingRight": 15,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#7AFA4C",
            "borderRadius": 50,
            "height": 10,
            "width": 10,
          }
        }
      />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#ffffff",
            "marginHorizontal": 20,
          }
        }
      >
        ± 50m
      </Text>
    </View>
  </View>
  
</View>
`;
